06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/PopsicleFinance/README.md

```diff
@@ -0,0 +1,29 @@
+(Did not have as much time to work on and organize this one)
+
+# Categories
+
+## Unit tests
+- ❌ Unit test for `deposit` - check that rewards and balances are updated
+    - ❓ Is there any way to check things about properties of `accounts` without making it public or using a harness? Would this be a good use of a `ghost`?
+- ✅ `deposit` increases balance
+- ✅ `withdraw` decreases balance
+- ✅ If user’s token balance increased, then `deposit` was called
+- ✅ If decreased, then `withdraw` was called
+- ❌ If someone deposits and then immediately collects fees, they won't earn any fees
+    - ❓ How can I test this one (assuming it's worth testing)? Their ETH balance will decrease by gas costs, so I can't check that their balance is unchanged. Checking that it's nonincreasing feels like it would miss the case when the fees earned are smaller than the gas costs.
+
+## Variable transitions
+- ❌ If `reward` increased, it was from a call to `deposit` or `withdraw`
+    - Need to set up a ghost to access `reward`
+- ❌ If caller's ETH balance increased, it was from a call to `collectFees()`
+    - ❓ This one is failing. It seems to think that a call to `withdraw` triggers an ETH transfer. See anything wrong with my spec `callersETHBalancesIncreasesOnlyFromCollectFees`?
+- ✅ ❓Create a ghost for the sum of all assets of users in the system. Confirm that only certain functions can change this amount.
+    - ❓I had to write a lot of ghosts for this one. The rule is `sumOfAllAssetsDoesntChange`, and it failed for `transfer` and `transferFrom` which I believe is expected because there's a bug in the Popsicle code. Would love a sanity check of my approach.
+
+## High level properties
+- ❌ Someone who never deposited should not be able to collect any fees
+    - ❓There is not a way to check something like "never deposited", right? And if I tried to use an invariant, I wouldn't be able to check something like "cannot collect fees" since the body of an invariant shouldn't involve state changes.
```

Comment:
I think you can do this in a rule by asserting that they don't have any balance in the contract and calling an arbitrary method and checking that their ether balance didn't increase.

Reply:
> by asserting that they don't have any balance

Do you mean `require`ing? That would make more sense to me

Reply:
Yes